rs/registry/canister/src/mutations/do_set_subnet_operational_level.rs

```diff
@@ -1,15 +1,229 @@
+use prost::Message;
+use std::collections::HashSet;
+
+pub mod operational_level {
+    pub const NORMAL: i32 = 1;
+    pub const DOWN_FOR_REPAIRS: i32 = 2;
+
+    pub const ALL_VALID_CODES: [i32; 2] = [NORMAL, DOWN_FOR_REPAIRS];
+
+    pub fn name(code: i32) -> Option<&'static str> {
+        let name = match code {
+            NORMAL => "normal",
+            DOWN_FOR_REPAIRS => "down_for_repairs",
+            _ => {
+                return None;
+            }
+        };
+
+        Some(name)
+    }
+
+    pub fn code(name: &str) -> Option<i32> {
+        let code = match name {
+            "normal" => NORMAL,
+            "down_for_repairs" => DOWN_FOR_REPAIRS,
+            _ => {
+                return None;
+            }
+        };
+
+        Some(code)
+    }
+}
 
 impl Registry {
-    pub fn do_set_subnet_operational_level(&mut self, _payload: SetSubnetOperationalLevelPayload) {
-        // TODO(NNS1-4224): Implement.
-        todo!();
+    pub fn do_set_subnet_operational_level(&mut self, payload: SetSubnetOperationalLevelPayload) {
+        println!("{LOG_PREFIX}do_set_subnet_operational_level: {payload:?}");
+        self.validate_set_subnet_operational_level(&payload);
+        let SetSubnetOperationalLevelPayload {
+            subnet_id,
+            operational_level,
+            ssh_readonly_access,
+            ssh_node_state_write_access,
+        } = payload;
+
+        let mut mutations: Vec<RegistryMutation> = vec![];
+
+        // Change SubnetRecord.
+        if let Some(subnet_id) = subnet_id {
+            mutations.push(modify_subnet_record_for_set_subnet_operational_level(
+                subnet_id,
+                self.get_subnet_or_panic(subnet_id),
+                operational_level,
+                ssh_readonly_access,
+            ));
+        }
+
+        // Change NodeRecord(s).
+        mutations.append(&mut modify_node_record_for_set_subnet_operational_level(
+            ssh_node_state_write_access,
+            |node_id| self.get_node_or_panic(node_id),
+        ));
+
+        self.maybe_apply_mutation_internal(mutations);
+    }
+
+    fn validate_set_subnet_operational_level(&self, payload: &SetSubnetOperationalLevelPayload) {
+        let SetSubnetOperationalLevelPayload {
+            subnet_id,
+            operational_level,
+            ssh_readonly_access,
+            ssh_node_state_write_access,
+        } = payload;
+
+        match subnet_id {
+            None => {
+                assert_eq!(
+                    operational_level, &None,
+                    "operational_level specified, but not subnet_id."
+                );
+                assert_eq!(
+                    ssh_readonly_access, &None,
+                    "ssh_readonly_access specified, but not subnet_id."
+                );
+            }
+
+            Some(_subnet_id) => {
+                // Nothing to do here.
+            }
+        }
+
+        validate_operational_level(*operational_level);
```

Comment:
The validation methods (including validate_set_subnet_operational_level) should return Result<(), String> and then you can do an expect at the top level.  The strategy of panicking in the Registry is not my favorite, but the decision to panic should be pushed up the stack.

Reply:
Done.

I agree that's better, but I was under the impression that for better or worse, in Registry, you just panic.

Now that I look more carefully, I see that we do not have a well-established pattern, but yes, there is precedent for what you are suggesting (hence, "Done.").